x-pack/legacy/plugins/apm/public/components/app/TransactionDetails/WaterfallWithSummmary/WaterfallContainer/Waterfall/waterfall_helpers/waterfall_helpers.ts

```diff
@@ -69,75 +56,150 @@ interface IWaterfallItemBase {
+  ITransaction,
+  'transaction'
+>;
+export type IWaterfallSpan = IWaterfallItemBase<ISpan, 'span'>;
+export type IWaterfallError = IWaterfallItemBase<IError, 'error'>;
+export type IWaterfallAgentMark = IWaterfallItemBase<IAgentMark, 'agentMark'>;
+
+export type IWaterfallItem =
+  | IWaterfallTransaction
+  | IWaterfallSpan
+  | IWaterfallError
+  | IWaterfallAgentMark;
 
 function getTransactionItem(
   transaction: Transaction,
   errorsPerTransaction: TraceAPIResponse['errorsPerTransaction']
-): IWaterfallItemTransaction {
+): IWaterfallTransaction {
   return {
-    id: transaction.transaction.id,
-    parentId: transaction.parent && transaction.parent.id,
-    serviceName: transaction.service.name,
-    name: transaction.transaction.name,
+    docType: 'transaction',
+    doc: {
+      id: transaction.transaction.id,
+      transaction,
+      errorsCount: errorsPerTransaction[transaction.transaction.id] || 0,
+      parentId: transaction.parent?.id,
+      serviceName: transaction.service.name,
+      timestamp: transaction.timestamp.us,
+      name: transaction.transaction.name
+    },
     duration: transaction.transaction.duration.us,
-    timestamp: transaction.timestamp.us,
     offset: 0,
-    skew: 0,
-    docType: 'transaction',
-    transaction,
-    errorCount: errorsPerTransaction[transaction.transaction.id] || 0
+    skew: 0
   };
 }
 
-function getSpanItem(span: Span): IWaterfallItemSpan {
+function getSpanItem(span: Span): IWaterfallSpan {
   return {
-    id: span.span.id,
-    parentId: span.parent && span.parent.id,
-    serviceName: span.service.name,
-    name: span.span.name,
+    docType: 'span',
+    doc: {
+      id: span.span.id,
+      span,
+      parentId: span.parent?.id,
+      serviceName: span.service.name,
+      timestamp: span.timestamp.us,
+      name: span.span.name
+    },
     duration: span.span.duration.us,
-    timestamp: span.timestamp.us,
+    offset: 0,
+    skew: 0
+  };
+}
+
+function getErrorItem(error: APMError): IWaterfallError {
+  return {
+    docType: 'error',
+    doc: {
+      id: error.error.id,
+      error,
+      message: error.error.log?.message || error.error.exception?.[0]?.message,
+      parentId: error.parent?.id,
+      serviceName: error.service.name,
+      timestamp: error.timestamp.us
+    },
     offset: 0,
     skew: 0,
-    docType: 'span',
-    span
+    duration: 0
   };
 }
 
+function getAgentMarks(transaction?: Transaction): IWaterfallAgentMark[] {
+  const agent = transaction?.transaction.marks?.agent;
+  if (!agent) {
+    return [];
+  }
+  return Object.entries(agent).map(([name, ms]) => ({
+    docType: 'agentMark',
+    doc: {
+      mark: name
+    },
+    offset: ms * 1000,
+    duration: 0,
+    skew: 0
+  }));
+}
+
 export function getClockSkew(
   item: IWaterfallItem,
   parentItem?: IWaterfallItem
 ) {
-  if (!parentItem) {
+  // parent should never be a type of agentMark
+  if (!parentItem || parentItem.docType === 'agentMark') {
```

Comment:
If parent is never of type `agentMark`, then why is this check necessary?

Reply:
Because parentItem is a type of IWaterfallItem, which can be an AgentMark, I know that will never happen, but typescript doesn't.

Reply:
But agentMarks should have the same interface as the other waterfall items. That's also what I mentioned elsewhere that I don't really understand `doc` vs `custom`.
I think `custom`  should contain things that `waterfall_helpers` will not care about and should not try to access. But the top level of the interface should contain the properties that waterfall_helpers do care about and will read from

Reply:
> But agentMarks should have the same interface as the other waterfall items

It does have.

> That's also what I mentioned elsewhere that I don't really understand `doc` vs `custom`

They are the same, I just thought that `item.doc` would be better than `item.custom`, because I also have a `docType`.